platform/zoul/dev/hp206c.c

```diff
@@ -0,0 +1,189 @@
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the Institute nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE INSTITUTE AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE INSTITUTE OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ *
+ * This file is part of the Contiki operating system.
+ *
+ */
+/*---------------------------------------------------------------------------*/
+/**
+ * \addtogroup zoul-sensor
+ * @{
+ *
+ * \file
+ *         HP206C Barometer and altimeter sensor driver
+ *
+ * \author
+ *         Javi Sanchez <jsanchez@zolertia.com>
+ */
+/*---------------------------------------------------------------------------*/
+#include <stdio.h>
+#include "contiki.h"
+#include "dev/i2c.h"
+#include "dev/hp206c.h"
+#include "lib/sensors.h"
+/*---------------------------------------------------------------------------*/
+#define DEBUG 0
+#if DEBUG
+#define PRINTF(...) printf(__VA_ARGS__)
+#else
+#define PRINTF(...)
+#endif
+/*---------------------------------------------------------------------------*/
+static uint8_t enabled;
+/*---------------------------------------------------------------------------*/
+static int
+status(int type)
+{
+  switch(type) {
+  case SENSORS_ACTIVE:
+  case SENSORS_READY:
+    return enabled;
+  }
+  return 0;
+}
+/*---------------------------------------------------------------------------*/
+static uint16_t
+hp206c_read_reg(uint8_t reg, uint8_t *buf, uint8_t num)
+{
+  if((buf == NULL) || (num <= 0)) {
+    return HP206C_ERROR;
+  }
+
+  i2c_master_enable();
+  if(i2c_single_send(HP206C_ADDR, reg) == I2C_MASTER_ERR_NONE) {
+    if(i2c_burst_receive(HP206C_ADDR, buf, num) == I2C_MASTER_ERR_NONE) {
+      return HP206C_SUCCESS;
+    }
+  }
+  return HP206C_ERROR;
+}
+/*---------------------------------------------------------------------------*/
+static int32_t
+hpc206c_convert(int32_t val)
+{
+  if((val & 0xF00000) == 0xF00000) {
+    val = (val & 0x0FFFFF) ^ 0xFFFFF;
+    return -val;
+  } else {
+    return val;
+  }
+  return HP206C_ERROR;
+}
```

Comment:
That converter function would be more efficient if it changes the value "in-place"; that is, it could be written as this:
```C
static void
hpc206c_convert(int32_t *val)
{
  if((*val & 0xF00000) == 0xF00000) {
    *val = -(*val ^ 0xFFFFFF);
  }
}
```